### PR TITLE
ilmerge nuget.build.tasks.pack

### PR DESCRIPTION
**Pull request by rohit21agrawal**
Fixes: https://github.com/NuGet/Home/issues/5011

CC: @emgarten @alpaix

### CODE REVIEW

File: src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/NuGet.Build.Tasks.Pack.Library.csproj [hunk 1]
```diff
@@ -5,11 +5,14 @@
     <TargetFrameworks>netstandard1.3;net45</TargetFrameworks>
     <NoWarn>$(NoWarn);CS1591</NoWarn>
     <AssemblyName>NuGet.Build.Tasks.Pack</AssemblyName>
-	<RootNamespace>$(AssemblyName)</RootNamespace>
-    <PackageId>NuGet.Build.Tasks.Pack.Library</PackageId>
+    <RootNamespace>$(AssemblyName)</RootNamespace>
+    <PackageId>NuGet.Build.Tasks.Pack</PackageId>
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/NuGet.Build.Tasks.Pack.Library.csproj]: Consider renaming the project to follow the package ID.

**Comment by alpaix** [comment]
What is the license for the new library?

**Comment by alpaix** [comment]
Does it merge symbols? What is the final size of the result assembly? Does it merge NuGet* assemblies only or all of the dependencies?

File: build/common.project.props [hunk 2]
```diff
@@ -23,6 +23,7 @@
     <SolutionPackagesFolder>$(RepositoryRootDirectory)packages\</SolutionPackagesFolder>
     <XunitConsoleExePath>$(SolutionPackagesFolder)xunit.runner.console.2.2.0\tools\xunit.console.x86.exe</XunitConsoleExePath>
     <ILMergeExePath>$(SolutionPackagesFolder)ILMerge.2.14.1208\tools\ILMerge.exe</ILMergeExePath>
+    <ILRepackExePath>$(SolutionPackagesFolder)ILRepack.2.0.13\tools\ILRepack.exe</ILRepackExePath>
```
Comment (alpaix) [inline build/common.project.props:26]: Package contains the props file defining the tool location. Also would it work if installed directly in target project?

**Review by alpaix — CHANGES_REQUESTED** [review]
Waiting for response.

File: src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/NuGet.Build.Tasks.Pack.Library.csproj (hunk 1 above)
Comment (emgarten) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/NuGet.Build.Tasks.Pack.Library.csproj]: @rohit21agrawal is the circular dependency problem fixed now that the SDK no longer depends on pack?

File: src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/NuGet.Build.Tasks.Pack.Library.csproj [hunk 3]
```diff
@@ -48,6 +51,25 @@
...
+  </Target>
+
+  <Target Name="ILRepackNuGetPack"  Condition="'$(TargetFramework)' != ''">
+    <Message Text="ILRepack: $(ILRepackExePath)" Importance="High" />
+    <ItemGroup>
+      <BuildArtifacts Include="$(OutputPath)*.dll" Exclude="$(OutputPath)*Pack.dll" />
+    </ItemGroup>
+    <PropertyGroup>
+      <PathToBuiltNuGetPack>$(OutputPath)NuGet.Build.Tasks.Pack.dll</PathToBuiltNuGetPack>
+      <IlmergeCommand>$(ILRepackExePath) $(PathToBuiltNuGetPack) /lib:$(OutputPath)$(RuntimeIdentifier)\publish @(BuildArtifacts, ' ') /out:$(OutputPath)NuGet.Build.Tasks.Pack.dll /internalize</IlmergeCommand>
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/NuGet.Build.Tasks.Pack.Library.csproj]: ``/parallel`` would help here

Comment (emgarten) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/NuGet.Build.Tasks.Pack.Library.csproj]: Is there a new .pdb created from this? If so it should get copied to the symbols folder with the nuget.exe merged one

File: src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/NuGet.Build.Tasks.Pack.Library.csproj (hunk 1 above)
Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/NuGet.Build.Tasks.Pack.Library.csproj]: yes.

File: build/common.project.props (hunk 2 above)
Comment (rohit21agrawal) [reply build/common.project.props:26]: @alpaix for some reason couldn't get it to work as a package reference.

File: src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/NuGet.Build.Tasks.Pack.Library.csproj (hunk 3 above)
Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/NuGet.Build.Tasks.Pack.Library.csproj]: we haven't been shipping symbols for this, as we don't create a symbols package.

**Comment by rohit21agrawal** [comment]
@alpaix the software is a part of Free Software Foundation and there is no license i could find on the project site.

the resulting dll for netstandard is 1.5 mb and for net45 is 2mb.

File: src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/NuGet.Build.Tasks.Pack.Library.csproj [hunk 4]
```diff
@@ -5,11 +5,13 @@
...
     <AssemblyName>NuGet.Build.Tasks.Pack</AssemblyName>
-	<RootNamespace>$(AssemblyName)</RootNamespace>
-    <PackageId>NuGet.Build.Tasks.Pack.Library</PackageId>
+    <RootNamespace>$(AssemblyName)</RootNamespace>
     <NetStandardImplicitPackageVersion Condition=" '$(TargetFramework)' == 'netstandard1.3' ">$(NetStandardPackageVersion)</NetStandardImplicitPackageVersion>
     <Shipping>true</Shipping>
     <PackProject>true</PackProject>
+    <NoBuild>true</NoBuild>
+    <AppendRuntimeIdentifierToOutputPath>false</AppendRuntimeIdentifierToOutputPath>
+    <RuntimeIdentifier>win10-x64</RuntimeIdentifier>
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/NuGet.Build.Tasks.Pack.Library.csproj:14]: Make sure this builds on Linux

**Review by emgarten — APPROVED** [review]

**Comment by alpaix** [comment]
@rohit21agrawal The license is here https://github.com/gluck/il-repack/blob/master/LICENSE.
It's licensed under the Apache License 2.0.

**Comment by alpaix** [comment]
@rohit21agrawal Are you going to pack both assemblies, netstandard and net45, into the final shipping package?

**Comment by rohit21agrawal** [comment]
@alpaix yes, both assemblies are going to be packed. also i have submitted request for approval of the ilrepack nuget package.

**Review by alpaix — APPROVED** [review]

**Comment by rohit21agrawal** [comment]
Closing in favour of https://github.com/NuGet/NuGet.Client/pull/1310

**Pull request closed without merging**